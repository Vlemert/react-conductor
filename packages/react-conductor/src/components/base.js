class Base {
  constructor(root, props) {
    this.root = root;
    this.props = props;
  }

  /**
   * Default event handler. Not yet implemented until we make events more
   * generic. For now handled in subclass.
   */
  handleEvent(event) {
    throw new Error('event handler not implemented');
  }

  /**
   * Definition of all (non-event) props that an element handles. Format:
   * { propKey: handlerFunction }
   * Should be overridden in subclass
   */
  get propHandlers() {
    return {};
  }

  /**
   * Definition of all event props that an element handles. Format:
   * { propKey: 'event-name' }
   * * Should be overridden in subclass
   */
  get propEvents() {
    return {};
  }

  /**
   * Helper that turns all entries in `propEvents` into actual event handlers
   */
  propEventHandlers = Object.entries(
    this.propEvents
  ).reduce((eventHandlers, [propKey, eventName]) => {
    eventHandlers[propKey] = this.handleEvent(eventName);
    return eventHandlers;
  }, {});

  /**
   * An object containing all props that an element handles, based on
   * `propHandlers` and `propEvents`.
   */
  allPropHandlers = {
    ...this.propHandlers,
    ...this.propEventHandlers
  };

  /**
   * First finalization pass. Here we check whether we'll need to do anything
   * in `commitMount` by checking if any of the handled props has a value.
   */
  finalizeInitialChildren(props) {
    let commitMount = false;

    Object.entries(props).forEach(([propKey, eventHandler]) => {
      if (this.allPropHandlers.hasOwnProperty(propKey)) {
        commitMount = true;
      }
    });

    return commitMount;
  }

  /**
   * Handle any props that were passed in the initial mount of the element. We
   * should probably register any events here as well, instead of in
   * `finalizeInitialChildren`
   */
  commitMount(props) {
    Object.entries(props).forEach(([propKey, propValue]) => {
      if (this.allPropHandlers.hasOwnProperty(propKey)) {
        this.allPropHandlers[propKey](propValue);
      }
    });
  }

  /**
   * Diff all props and check whether we're handling them in this element
   * instance. If a handled prop was changed, return the change as
   * `updatePayload` so `commitUpdate` can work with the changes.
   *
   * `updatePayload` will be a list of tuples: [[propKey, newValue]]
   */
  prepareUpdate(oldProps, newProps) {
    const updatePayload = [];
    const mergedProps = {};

    Object.entries(oldProps).forEach(([key, value]) => {
      if (!this.allPropHandlers.hasOwnProperty(key)) {
        return;
      }

      mergedProps[key] = [value, undefined];
    });

    Object.entries(newProps).forEach(([key, value]) => {
      if (!this.allPropHandlers.hasOwnProperty(key)) {
        return;
      }

      if (mergedProps[key]) {
        mergedProps[key][1] = value;
        return;
      }

      mergedProps[key] = [undefined, value];
    });

    Object.entries(mergedProps).forEach(([propKey, [oldValue, newValue]]) => {
      if (oldValue !== newValue) {
        updatePayload.push([propKey, newValue]);
      }
    });

    return updatePayload.length ? updatePayload : null;
  }

  /**
   * Handle any props (that this element handles) that changed since the last
   * pass. See `prepareUpdate` for the origin / shape of `updatePayload`.
   */
  commitUpdate(updatePayload, oldProps, newProps) {
    updatePayload.forEach(([propKey, value]) => {
      if (this.allPropHandlers.hasOwnProperty(propKey)) {
        this.allPropHandlers[propKey](value);
      }
    });
  }
}

export default Base;
