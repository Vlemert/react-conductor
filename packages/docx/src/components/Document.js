class Document {
  // Stores all the children
  children = [];

  constructor(root, props) {
    this.root = root;
    this.props = props;

    // docx instance for adding text nodes (Note - This text nodes are different when compared to DOM)
    this.adder = this.root.doc.createP();
  }

  appendChild(child) {
    this.children.push(child);
  }

  removeChild(child) {
    const index = this.children.indexOf(child);
    this.children.slice(index, 1);
  }

  renderChildNode() {
    for (let i = 0; i < this.children.length; i += 1) {
      if (typeof this.children[i] === 'string') {
        this.adder.addText(this.children[i]);
      } else if (typeof this.children[i] === 'object') {
        this.children[i].render();
      }
    }
  }

  render() {
    this.renderChildNode();
  }
}

export default Document;
