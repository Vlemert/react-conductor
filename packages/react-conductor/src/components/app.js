import Base from './base';
import Window from './window';

class App extends Base {
  constructor(root, props) {
    super(root, props);

    this.childWindows = new Set();
  }

  appendChild(child) {
    if (child instanceof Window) {
      this.childWindows.add(child);
    }
  }

  removeChild(child) {
    if (child instanceof Window) {
      this.childWindows.delete(child);
    }
  }

  // This is mainly here to test whether I could get this to work
  // Might want to move all the app.dock stuff to a different element
  handleDockBounce(newValue) {
    if (newValue) {
      this.dockBounceId = this.root.electronApp.dock.bounce(
        newValue === true ? undefined : newValue
      );
    } else {
      if (this.dockBounceId !== undefined) {
        this.root.electronApp.dock.cancelBounce(this.dockBounceId);
      }
    }
  }

  handleEvent(event) {
    // TODO: make event handling more generic + remove old handlers on change
    return (newValue, oldValue) => {
      this.root.electronApp.on(event, newValue);
    };
  }

  get propEvents() {
    return {
      onWillFinishLaunching: 'will-finish-launching',
      // Will not trigger because we wait with rendering until this event has
      // passed. Should probably call this in commitMount if set
      onReady: 'ready',
      onWindowAllClosed: 'window-all-closed',
      onBeforeQuit: 'before-quit',
      onWillQuit: 'will-quit',
      onQuit: 'quit',
      // open-file needs to be registered very early, should probably do that in
      // the constructor and call the prop/handler in commitMount
      onOpenFile: 'open-file',
      // Same as open-file
      onOpenUrl: 'open-url',
      onActivate: 'activate',
      onContinueActivity: 'continue-activity',
      onNewWindowForTab: 'new-window-for-tab',
      onBrowserWindowBlur: 'browser-window-blur',
      onBrowserWindowFocus: 'browser-window-focus',
      onBrowserWindowCreated: 'browser-window-created',
      onWebContentsCreated: 'web-contents-created',
      onCertificateError: 'certificate-error',
      onSelectClientCertificate: 'select-client-certificate',
      onLogin: 'login',
      onGpuProcessCrashed: 'gpu-process-crashed',
      onAccessibilitySupportChanged: 'accessibility-support-changed'
    };
  }

  /**
   * Note: we need to bind the handlers to preserve `this` context.
   */
  get propHandlers() {
    return {
      dockBounce: this.handleDockBounce.bind(this)
    };
  }
}

export default App;
