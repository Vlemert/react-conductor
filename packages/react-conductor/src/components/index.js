import Root from './root';
import App from './app';
import Window from './window';
import Menu from './menu';

export { Root, App, Window, Menu };
