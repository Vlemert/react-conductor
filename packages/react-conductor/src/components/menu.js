import { Menu as ElectronMenu } from 'electron';

import createEventManager from '../utils/create-event-manager';
import Base from './base';

/**
 * Menus are an interesting concept in Electron. You can create as many
 * instances of menus as you want, and they won't do anything until you either
 * set them as an applicationMenu or pop them up as a context menu.
 *
 * I think we can leverage this, by making their behaviour depend on where they
 * are rendered. e.g. we could make a menu inside a window the applicationMenu
 * when the window gets focus, or do something smart with context menus.
 *
 * For now, we'll start with a basic implementation.
 */
class Menu extends Base {
  constructor(props) {
    super(props);

    this.menu = new ElectronMenu();
    this.eventManager = createEventManager(this.menu);
  }

  getPublicInstance() {
    // TODO: think about whether we want to grant the user full access here
    return this.menu;
  }

  get propEvents() {
    return {};
  }

  /**
   * Note: we need to bind the handlers to preserve `this` context.
   */
  get propHandlers() {
    return {};
  }
}

export default Menu;
