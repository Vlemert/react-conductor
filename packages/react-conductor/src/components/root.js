import App from './app';

class Root {
  appendChild(child) {
    // Not sure if we want to enforce this, but don't want to deal with the
    // repercussions of not enforcing this right now. (windows being children
    // of root etc)
    if (!(child instanceof App)) {
      throw new Error('Root element should be an App!');
    }

    this.appElement = child;
  }

  removeChild(child) {
    this.appElement = null;
  }
}

export default Root;
