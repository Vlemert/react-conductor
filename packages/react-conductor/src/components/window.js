import { BrowserWindow } from 'electron';

import createEventManager from '../utils/create-event-manager';
import Base from './base';

class Window extends Base {
  constructor(props) {
    super(props);

    this.childWindows = new Set();

    this.browserWindow = new BrowserWindow({
      show: !!props.show
    });
    this.eventManager = createEventManager(this.browserWindow);
  }

  appendChild(child) {
    if (child instanceof Window) {
      this.childWindows.add(child);
    }
  }

  removeChild(child) {
    if (child instanceof Window) {
      this.childWindows.delete(child);
    }
  }

  getPublicInstance() {
    // TODO: think about whether we want to grant the user full access here
    return this.browserWindow;
  }

  handleEvent(event) {
    // TODO: can we move this to base?
    return newValue => {
      this.eventManager(event, newValue);
    };
  }

  /**
   * This handler will make sure the `defaultSize`, `size`, and `onResize` props
   * are handled correctly. It will
   *
   * We can probably improve this code by looking at the implementation in the
   * react codebase (ReactDOMFiberInput.js for example).
   */
  handleSize(propValue, newProps, oldProps) {
    // TODO: fix this check, only do it in dev (need to setup environments)
    // if (oldProps && newProps.defaultSize !== oldProps.defaultSize) {
    //   console.log(newProps.defaultSize, oldProps.defaultSize);
    //   console.warn('You cannot change the value of `defaultSize` after mount');
    // }

    if (
      (!oldProps && newProps.onResize) ||
      (oldProps && newProps.onResize !== oldProps.onResize)
    ) {
      this.eventManager('resize', () => {
        newProps.onResize(this.browserWindow.getSize());
      });
    }

    if (!newProps.defaultSize && !newProps.size) {
      this.browserWindow.setResizable(true);
      return;
    }

    // TODO: only do this on mount?
    if (newProps.defaultSize) {
      this.browserWindow.setSize(...newProps.defaultSize);
      this.browserWindow.setResizable(true);
      return;
    }

    if (newProps.size && !newProps.onResize) {
      // TODO: reactdom would warn if you do this, something like:
      // "If the field should be mutable use `defaultValue`. Otherwise, set
      // either `onChange` or `readOnly`."
      this.browserWindow.setSize(...newProps.size);
      this.browserWindow.setResizable(false);
      return;
    }

    if (newProps.size && newProps.onResize) {
      this.browserWindow.setSize(...newProps.size);
      this.browserWindow.setResizable(true);
    }
  }

  /**
   * Same as `handleSize` above. This is a lot of practically the same code, we
   * might want to do something with that in the future. Lets wait with that
   * until the implementation is complete (including error/warning messages).
   */
  handlePosition(propValue, newProps, oldProps) {
    if (
      (!oldProps && newProps.onMove) ||
      (oldProps && newProps.onMove !== oldProps.onMove)
    ) {
      this.eventManager('move', () => {
        newProps.onMove(this.browserWindow.getPosition());
      });
    }

    if (!newProps.defaultPosition && !newProps.position) {
      this.browserWindow.setMovable(true);
      return;
    }

    // TODO: only do this on mount?
    if (newProps.defaultPosition) {
      this.browserWindow.setPosition(...newProps.defaultPosition);
      this.browserWindow.setMovable(true);
      return;
    }

    if (newProps.position && !newProps.onMove) {
      // TODO: reactdom would warn if you do this, something like:
      // "If the field should be mutable use `defaultValue`. Otherwise, set
      // either `onChange` or `readOnly`."
      this.browserWindow.setPosition(...newProps.position);
      this.browserWindow.setMovable(false);
      return;
    }

    if (newProps.position && newProps.onMove) {
      this.browserWindow.setPosition(...newProps.position);
      this.browserWindow.setMovable(true);
    }
  }

  /**
   * Set a path for the window to load. If the path doesn't start with either
   * http://, https://, or file://, we assume it's a local path and prepend it
   * with file://
   */
  handlePath(path) {
    if (
      path.indexOf('file://') === 0 ||
      path.indexOf('http://') === 0 ||
      path.indexOf('https://') === 0
    ) {
      this.browserWindow.loadURL(path);
      return;
    }

    this.browserWindow.loadURL(`file://${path}`);
  }

  /**
   * Note: we need to bind the handlers to preserve `this` context.
   */
  get propHandlers() {
    const handleSize = this.handleSize.bind(this);
    const handlePosition = this.handlePosition.bind(this);
    return {
      // We can bind multiple props to the same handler because the base class
      // will only call each handler once when props change. This does mean our
      // handler needs to look at all the props when it's called.
      defaultSize: handleSize,
      size: handleSize,
      // I think it would be nice if this could be handled like other events.
      // That would mean event handling needs some kind of mapping support. (to
      // map the raw event to something 'nice' for the user)
      onResize: handleSize,
      // We handle position in the same way as we do with size
      defaultPosition: handlePosition,
      position: handlePosition,
      onMove: handlePosition,
      path: this.handlePath.bind(this)
    };
  }
}

export default Window;
