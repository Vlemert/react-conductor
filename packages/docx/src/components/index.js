import WordDocument from './WordDocument';
import Text from './Text';
import Document from './Document';

export { WordDocument, Text, Document };
