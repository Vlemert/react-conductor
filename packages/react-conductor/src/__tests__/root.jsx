import React from 'react';
import { app, BrowserWindow } from 'electron';

import { Root } from '../components';
import { Window } from '../index';
import testRender from '../test-renderer';

jest.mock('electron');

describe('Root', () => {
  beforeEach(() => {
    BrowserWindow.mockReset();
  });

  test('is rendered as the root of the element tree', () => {
    const Application = () => null;

    const wrapper = testRender(<Application />);

    expect(wrapper).toBeInstanceOf(Root);
  });

  test('throws if something else than an App is appended to it', () => {
    // Hide react error details from the console
    const originalConsoleError = console.error;
    console.error = jest.fn();

    const Application = () => <Window />;

    expect(() => testRender(<Application />)).toThrowErrorMatchingSnapshot();

    // And restore console.error behaviour
    console.error = originalConsoleError;
  });

  test('throws when trying to render a string', () => {
    // Hide react error details from the console
    const originalConsoleError = console.error;
    console.error = jest.fn();

    const Application = () => 'hello';

    expect(() => testRender(<Application />)).toThrowErrorMatchingSnapshot();

    // And restore console.error behaviour
    console.error = originalConsoleError;
  });

  test('throws when trying to render an invalid component', () => {
    // Hide react error details from the console
    const originalConsoleError = console.error;
    console.error = jest.fn();

    const Fake = 'fake';
    const Application = () => <Fake />;

    expect(() => testRender(<Application />)).toThrowErrorMatchingSnapshot();

    // And restore console.error behaviour
    console.error = originalConsoleError;
  });
});
