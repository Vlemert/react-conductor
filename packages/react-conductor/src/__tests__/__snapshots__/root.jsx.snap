// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root throws if something else than an App is appended to it 1`] = `"Root element should be an App!"`;

exports[`Root throws when trying to render a string 1`] = `"TextElements are not supported! (do you have some text in your JSX?)"`;

exports[`Root throws when trying to render an invalid component 1`] = `"type 'fake' is not implemented (yet?)!"`;
